File: javascript/src/ParameterTypeRegistry.ts
import CucumberExpressionError from './CucumberExpressionError.js'
import CucumberExpressionGenerator from './CucumberExpressionGenerator.js'
import defineDefaultParameterTypes from './defineDefaultParameterTypes.js'
import { AmbiguousParameterTypeError } from './Errors.js'
import ParameterType from './ParameterType.js'
import { DefinesParameterType } from './types.js'

export default class ParameterTypeRegistry implements DefinesParameterType {
  private readonly parameterTypeByName = new Map<string, ParameterType<unknown>>()
  private readonly parameterTypesByRegexp = new Map<string, Array<ParameterType<unknown>>>()

  constructor() {
    defineDefaultParameterTypes(this)
  }

  get parameterTypes(): IterableIterator<ParameterType<unknown>> {
    return this.parameterTypeByName.values()
  }

  public lookupByTypeName(typeName: string) {
    return this.parameterTypeByName.get(typeName)
  }

  public lookupByRegexp(
    parameterTypeRegexp: string,
    expressionRegexp: RegExp,
    text: string
  ): ParameterType<unknown> | undefined {
    const parameterTypes = this.parameterTypesByRegexp.get(parameterTypeRegexp)
    if (!parameterTypes) {
      return undefined
    }
    if (parameterTypes.length > 1 && !parameterTypes[0].preferForRegexpMatch) {
      // We don't do this check on insertion because we only want to restrict
      // ambiguity when we look up by Regexp. Users of CucumberExpression should
      // not be restricted.
      const generatedExpressions = new CucumberExpressionGenerator(
        () => this.parameterTypes
      ).generateExpressions(text)
      throw AmbiguousParameterTypeError.forRegExp(
        parameterTypeRegexp,
        expressionRegexp,
        parameterTypes,
        generatedExpressions
      )
    }
    return parameterTypes[0]
  }

  public defineParameterType(parameterType: ParameterType<unknown>) {
    if (parameterType.name !== undefined) {
      if (this.parameterTypeByName.has(parameterType.name)) {
        if (parameterType.name.length === 0) {
          throw new CucumberExpressionError(`The anonymous parameter type has already been defined`)
        } else {
          throw new CucumberExpressionError(
            `There is already a parameter type with name ${parameterType.name}`
          )
        }
      }
      this.parameterTypeByName.set(parameterType.name, parameterType)
    }

    for (const parameterTypeRegexp of parameterType.regexpStrings) {
      if (!this.parameterTypesByRegexp.has(parameterTypeRegexp)) {
        this.parameterTypesByRegexp.set(parameterTypeRegexp, [])
      }
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      const parameterTypes = this.parameterTypesByRegexp.get(parameterTypeRegexp)!
      const existingParameterType = parameterTypes[0]
      if (
        parameterTypes.length > 0 &&
        existingParameterType.preferForRegexpMatch &&
        parameterType.preferForRegexpMatch
      ) {
        throw new CucumberExpressionError(
          'There can only be one preferential parameter type per regexp. ' +
            `The regexp /${parameterTypeRegexp}/ is used for two preferential parameter types, {${existingParameterType.name}} and {${parameterType.name}}`
        )
      }
      if (parameterTypes.indexOf(parameterType) === -1) {
        parameterTypes.push(parameterType)
        this.parameterTypesByRegexp.set(
          parameterTypeRegexp,
          parameterTypes.sort(ParameterType.compare)
        )
      }
    }
  }
}

File: dart/lib/src/parameter_type_registry.dart
import 'package:cucumber_expressions/src/define_default_parameter_types.dart';
import 'package:cucumber_expressions/src/errors.dart';
import 'package:cucumber_expressions/src/parameter_type.dart';

/// A registry of [ParameterType]s, seeded with the built-in types and used to
/// resolve parameter references in expressions.
class ParameterTypeRegistry {
  /// Creates a registry populated with the default parameter types.
  ParameterTypeRegistry() {
    defineDefaultParameterTypes(defineParameterType);
  }

  final Map<String, ParameterType<Object?>> _parameterTypeByName = {};
  final Map<String, List<ParameterType<Object?>>> _parameterTypesByRegexp = {};

  /// Registers [parameterType] for use in expressions.
  ///
  /// Throws [CucumberExpressionException] if its name is already registered or
  /// if it conflicts with another preferential type for the same regexp.
  void defineParameterType<T>(ParameterType<T> parameterType) {
    final name = parameterType.name;
    if (name != null) {
      if (_parameterTypeByName.containsKey(name)) {
        if (name.isEmpty) {
          throw CucumberExpressionException(
            'The anonymous parameter type has already been defined',
          );
        } else {
          throw CucumberExpressionException(
            'There is already a parameter type with name $name',
          );
        }
      }
      _parameterTypeByName[name] = parameterType;
    }

    for (final parameterTypeRegexp in parameterType.regexpStrings) {
      final parameterTypes = _parameterTypesByRegexp.putIfAbsent(
        parameterTypeRegexp,
        () => <ParameterType<Object?>>[],
      );
      final existingParameterType =
          parameterTypes.isNotEmpty ? parameterTypes[0] : null;
      if (parameterTypes.isNotEmpty &&
          existingParameterType!.preferForRegexpMatch &&
          parameterType.preferForRegexpMatch) {
        throw CucumberExpressionException(
          'There can only be one preferential parameter type per regexp. '
          'The regexp /$parameterTypeRegexp/ is used for two preferential '
          'parameter types, {${existingParameterType.name}} and '
          '{${parameterType.name}}',
        );
      }
      if (!parameterTypes.contains(parameterType)) {
        parameterTypes
          ..add(parameterType)
          ..sort(compareParameterTypes);
      }
    }
  }
}

Iterable<ParameterType<Object?>> registeredParameterTypes(
  ParameterTypeRegistry registry,
) {
  return registry._parameterTypeByName.values;
}

ParameterType<Object?>? registeredParameterTypeByName(
  ParameterTypeRegistry registry,
  String typeName,
) {
  return registry._parameterTypeByName[typeName];
}

List<ParameterType<Object?>>? registeredParameterTypesByRegexp(
  ParameterTypeRegistry registry,
  String parameterTypeRegexp,
) {
  return registry._parameterTypesByRegexp[parameterTypeRegexp];
}
